### ISSUE TITLE
OBO Importer generates errors when attempting to load Sequence Ontology

### ISSUE THREAD

**Issue by risharde** [tripal/tripal#1209]
* Tripal Version: 3.6
* Drupal Version: 7
* PostgreSQL Version: 9.6
* PHP Version: 7.1.27

ISSUE:
OBO Loader fails to import Sequence Ontology and outputs error (attached below)

STEPS TO REPRODUCE:
Create a tripal 3 docker container (https://github.com/statonlab/docker-containers) which has Tripal V3.1
Edit /var/lib/pgsql/9.6/data/postgres.conf change max_locks_per_transaction to 128
Go to sites/all/modules/custom/ and delete tripal folder
git clone https://github.com/tripal/tripal.git
drush updatedb
Go to Tripal -> Data Loaders -> Chado Vocabularies -> OBO Vocabulary Loader -> Select Sequence Ontology

```
Please update the database using "drush updatedb" before continuing                                          [status]

2021-05-31 03:36:01
Tripal Job Launcher
Running as user 'admin'
-------------------
2021-05-31 03:36:01: There are 1 jobs queued.
2021-05-31 03:36:01: Job ID 72.
2021-05-31 03:36:01: Calling: tripal_run_importer(35)

Running 'OBO Vocabulary Loader' importer
NOTE: Loading of file is performed using a database transaction.
If it fails or is terminated prematurely then all insertions and
updates are rolled back and will not be found in the database

[site http://default] [TRIPAL ERROR] [TRIPAL_CHADO] chado_select_record: the foreign key definition for 'db_id' on table 'dbxref' returned no results where the definition supplied was Array(    [name] => NCIT)
[site http://default] [TRIPAL ERROR] [TRIPAL_CHADO] chado_select_record: the foreign key definition for 'dbxref_id' on table 'cvterm' returned no results where the definition supplied was Array(    [db_id] => Array        (            [name] => NCIT        )    [accession] => C25693)
[site http://default] [TRIPAL ERROR] [TRIPAL_CHADO] chado_select_record: the foreign key definition for 'db_id' on table 'dbxref' returned no results where the definition supplied was Array(    [name] => rdfs)
[site http://default] [TRIPAL ERROR] [TRIPAL_CHADO] chado_select_record: the foreign key definition for 'dbxref_id' on table 'cvterm' returned no results where the definition supplied was Array(    [db_id] => Array        (            [name] => rdfs        )    [accession] => comment)
Downloading URL http://purl.obolibrary.org/obo/so.obo, saving to /tmp/obo_cBND6m
Step 1: Preloading File /tmp/obo_cBND6m...
Found the following namespaces: sequence.,544 bytes.
Step 2: Examining relationships...
Step 3: Loading type defs...: 49,616,664  bytes.
Cannot add subset, "SOFA" to term: adjacent_to. ERROR: ChadoRecord::insert(). Could not insert a record into the table, cvtermprop, with the following values: Array
(
    [cvterm_id] => 12
    [type_id] =>
    [value] => SOFA
)
. ERROR: SQLSTATE[23502]: Not null violation: 7 ERROR:  null value in column "type_id" violates not-null constraint
DETAIL:  Failing row contains (2, 12, null, SOFA, 0)..
[site http://default] [TRIPAL ERROR] [TRIPAL_JOB] Cannot add subset, "SOFA" to term: adjacent_to. ERROR: ChadoRecord::insert(). Could not insert a record into the table, cvtermprop, with the following values: Array(    [cvterm_id] => 12    [type_id] =>     [value] => SOFA). ERROR: SQLSTATE[23502]: Not null violation: 7 ERROR:  null value in column "type_id" violates not-null constraintDETAIL:  Failing row contains (2, 12, null, SOFA, 0)..
```

**Comment by risharde** [comment]
I can also confirm that Tripal V3.4 does not suffer from this issue. I will be testing V3.5 to see whether it also suffers from the issue.

**Comment by risharde** [comment]
Tripal V3.5 error when performing OBO Loader (Sequence Ontology)

```
Please update the database using "drush updatedb" before continuing                                          [status]

2021-05-31 04:40:25
Tripal Job Launcher
Running as user 'admin'
-------------------
2021-05-31 04:40:25: There are 1 jobs queued.
2021-05-31 04:40:25: Job ID 73.
2021-05-31 04:40:25: Calling: tripal_run_importer(35)

Running 'OBO Vocabulary Loader' importer
NOTE: Loading of file is performed using a database transaction.
If it fails or is terminated prematurely then all insertions and
updates are rolled back and will not be found in the database

[site http://default] [TRIPAL ERROR] [TRIPAL_CHADO] chado_select_record: the foreign key definition for 'db_id' on table 'dbxref' returned no results where the definition supplied was Array(    [name] => NCIT)
[site http://default] [TRIPAL ERROR] [TRIPAL_CHADO] chado_select_record: the foreign key definition for 'dbxref_id' on table 'cvterm' returned no results where the definition supplied was Array(    [db_id] => Array        (            [name] => NCIT        )    [accession] => C25693)
[site http://default] [TRIPAL ERROR] [TRIPAL_CHADO] chado_select_record: the foreign key definition for 'db_id' on table 'dbxref' returned no results where the definition supplied was Array(    [name] => rdfs)
[site http://default] [TRIPAL ERROR] [TRIPAL_CHADO] chado_select_record: the foreign key definition for 'dbxref_id' on table 'cvterm' returned no results where the definition supplied was Array(    [db_id] => Array        (            [name] => rdfs        )    [accession] => comment)
Downloading URL http://purl.obolibrary.org/obo/so.obo, saving to /tmp/obo_rZxp5i
Step 1: Preloading File /tmp/obo_rZxp5i...
Found the following namespaces: sequence.,976 bytes.
Step 2: Examining relationships...
Step 3: Loading type defs...: 48,873,096  bytes.
Cannot add subset, "SOFA" to term: adjacent_to. ERROR: ChadoRecord::insert(). Could not insert a record into the table, cvtermprop, with the following values: Array
(
    [cvterm_id] => 7
    [type_id] =>
    [value] => SOFA
)
. ERROR: SQLSTATE[23502]: Not null violation: 7 ERROR:  null value in column "type_id" violates not-null constraint
DETAIL:  Failing row contains (1, 7, null, SOFA, 0)..
[site http://default] [TRIPAL ERROR] [TRIPAL_JOB] Cannot add subset, "SOFA" to term: adjacent_to. ERROR: ChadoRecord::insert(). Could not insert a record into the table, cvtermprop, with the following values: Array(    [cvterm_id] => 7    [type_id] =>     [value] => SOFA). ERROR: SQLSTATE[23502]: Not null violation: 7 ERROR:  null value in column "type_id" violates not-null constraintDETAIL:  Failing row contains (1, 7, null, SOFA, 0)..
```

**Comment by spficklin** [comment]
I just ran the OBO loader on my local machine with Tripal v3.6 (dev branch) and it loaded just fine.  It only took 21 seconds (see report below). So something is different between your site and mine.  

Is your site pulling the sequence ontology from this URL:  http://purl.obolibrary.org/obo/so.obo

```
2021-06-01 09:28:05
Tripal Job Launcher
Running as user 'administrator'
-------------------
2021-06-01 09:28:05: There are 4 jobs queued.
2021-06-01 09:28:05: Job ID 8.
2021-06-01 09:28:05: Calling: tripal_tripal_cron_notification()
2021-06-01 09:28:07: Job ID 9.
2021-06-01 09:28:07: Calling: tripal_expire_collections()
2021-06-01 09:28:07: Job ID 10.
2021-06-01 09:28:07: Calling: tripal_expire_files()
2021-06-01 09:28:07: Job ID 11.
2021-06-01 09:28:07: Calling: tripal_run_importer(8)

Running 'OBO Vocabulary Loader' importer
NOTE: Loading of file is performed using a database transaction. 
If it fails or is terminated prematurely then all insertions and 
updates are rolled back and will not be found in the database

Downloading URL http://purl.obolibrary.org/obo/so.obo, saving to /tmp/obo_SX79C4
Step 1: Preloading File /tmp/obo_SX79C4...
Found the following namespaces: sequence.,552 bytes.
Step 2: Examining relationships...
Step 3: Loading type defs...: 50,158,696  bytes.
Step 4: Loading terms... %. Memory: 50,164,984 bytes.
Step 5: Cleanup...100.00 %. Memory: 50,408,248 bytes.

Done.
Committing Transaction...

Updating the cv_root_mview materialized view...
Done.
Updating the db2cv_mview materialized view...
Done.
Updating cvtermpath table. This may take a while...
- Loading paths for vocabulary: sequence
Building cvterm paths for vocabulary: sequence
Clearing the cvtermpath table for this vocabulary...
Clearing completed.
Retrieving relationships...
Found 865 relationships in this vocabulary.
Note: Progress updates occur as each term is processed and some terms take longer than others.
Processing terms...
Processing tree root: sequence_variant...
Processing tree root: sequence_feature...bytes. Duration: 0.00 mins
Percent complete: 15%. Memory: 39,375,368 bytes. Duration: 0.03 mins
Processing tree root: sequence_collection...tes. Duration: 0.07 mins
Processing tree root: sequence_attribute...tes.
Remapping Chado Controlled vocabularies to Tripal Terms... 0.08 mins
Done.

ficklin@Thor /var/www/html/sites/all/modules $ date
Tue 01 Jun 2021 09:28:26 AM PDT
```

**Comment by risharde** [comment]
@spficklin thanks for checking on this! To confirm, yes: http://purl.obolibrary.org/obo/so.obo (the default) was tested.

**Comment by srobb1** [comment]
Hello All. I was getting errors with the loader as well when I was loaded plana.owl with Tripal 3, v 7.x-3.7.

This is property in my ontology that is referenced but not a term belonging to PLANA. Also this term is a property and not a term, and this term has a weirdness that it does not have any term that says it is the source term.

```
Performing EBI OLS Lookup for: BFO:0000062
Error: no data with http://www.ebi.ac.uk/ols/api/ontologies/bfo/terms/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FBFO_0000062.
```

i got this error as well, it is compounding the RO and BPSP properties.
```
Performing EBI OLS Lookup for: RO:0001001
Error: no data with http://www.ebi.ac.uk/ols/api/ontologies/ro/terms/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FBFO_00000500001001. 
```

I made some changes to the loader, and i can get theses to load by searching for the terms to find out if they are terms or properties and which ontology they belong to, then retrieving them with the correct api call.

Here are my fixes. My ontology now loads. I don't know if these changes will help with the above errors.
#1228 


Sofia

**Comment by dsenalik** [comment]
I am looking at old issues, I tried this and there now appears to be a PHP8 issue
```TypeError: count(): Argument #1 ($value) must be of type Countable|array, string given in form_set_error() (line 1660 of /var/www/prova/drupal-7.92/includes/form.inc).```

I wonder if this only showed up under PHP8? This is due to a typo in the form validation error message, and the error is caused by an incorrect checking of file existence when no file has been specified, that is, when the "Local File" box is left empty.

Fix the ```form_set_error``` typo, and skip check if ```$uobo_file``` is not set in pull #1443

**Comment by dsenalik** [comment]
I see this bug got copied over to Tripal 4, we need to fix it there too
Issue #1444

**Comment by dsenalik** [comment]
Fixed with merged PR #1443
